File: 2131-longest-palindrome-by-concatenating-two-letter-words/main.py
class Solution:
    def longestPalindrome(self, words: List[str]) -> int:

        freq = defaultdict(int)
        ans = f = 0

        for word in words:
            freq[word] += 1

        
        for k, v in freq.items():
            x, y = k
            if x == y:
                if v & 1 and not f: f = 2
                ans += 2 * (v//2)
            elif y+x in freq:
                ans += min(freq[k], freq[y+x])

        return 2 * ans + f        

File: 2131-longest-palindrome-by-concatenating-two-letter-words/main.go
func longestPalindrome(words []string) int {
    freq := make(map[string]int)
    ans := 0
    f := 0

    for _, word := range words {
        freq[word]++
    }

    for k, v := range freq {
        x, y := string(k[0]), string(k[len(k)-1])
        if x == y {
            if v & 1 != 0 && f == 0 {
                f = 2
            }
            ans += 2 * (v / 2)
        } else if _, ok := freq[y+x]; ok {
            ans += min(v, freq[y+x])
        }
    }

    return 2*ans + f
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}
